---
title: "PICS LOG precision and recall"
output: html_document
author: Max Dippel
date: "2023-11-15"
---

The goal of this rmd is to get precision and recall heat maps to try and develop an imperical method to classify emvars. 

Precision is the fraction of relevant instances divided by all retrieved instances. This is also called the positive predictive value. 
Precision = True Positives / (True Positives + False Positives)

Recall is the fraction of all retrieved instances divided by all the retrieved instances. This is also called the sensitivity. 
Recall = True Positives / (True Positives + False Negatives)

This html will give precision and recall plots for these parameters: 

1. Skew

2. Max Allele Fold Change

3. Fold Change p-value with Benjamhini Hochberg correction

4. Fold Change p-value with Bonferroni correction

To calculate the true_positives and false_positives we use these thresholds:

True Positive: max_pics>=0.5

FalsePositive: max_pics<=0.001

```{r, echo=FALSE, message=FALSE}
# loading packages
library(tidyverse)
library(RColorBrewer)
library(rmarkdown)
```

# HTML version

This is the place in the html you can tell which mpra data set and specific filters are being used. In the Rmd there are multiple options and you need to delete ", eval=FALSE, echo=FALSE" from the R chunk you want to use. This will make that R chunk be used in the making of the html and display that code chunk in the html. To create the html with the correct mpra loading code chunk, you need to do to run the render chunk which is directly below it. Keep the render chunk " eval=FALSE, Echo=FALSE" so it does not appear in the html. 

```{r, eval=FALSE, echo=FALSE}
# This html version uses the mpra data from 20230825_mpra_merge_Tcell_glm_john
# This html version has no filter 
mpra.dir <- "/nfs/jray/screens/ALL_MPRAs/Ho_et_al_analysis/Downstream_Analysis/20230825_stimulated_primary_Tcell_glm_replication/john_filter/data/20230825_mpra_merge_Tcell_glm_john.txt"
mpra <- read.table(mpra.dir, sep="\t", header=T)
mpra_name <- "20230825_mpra_merge_Tcell_glm_john"
filter <- "none"
```

```{r, eval=FALSE, echo=FALSE}
# These are the different options for saving the html. They depend on the values mpra_name and filter.
# This is going to be the default 
if(filter == "none" & mpra_name == "20230825_mpra_merge_Tcell_glm_john"){
  render("20231115_LOG_precision_recall_analysis.Rmd",output_file = paste0(mpra_name, "_LOG_PICS_precision_recall",".html"))
}
```

```{r, eval=FALSE, echo=FALSE}
# This html version uses the mpra data from 20230518_unstimulated_jurkat_glm
# This html version has no filter 
mpra.dir <- "/nfs/jray/screens/ALL_MPRAs/Ho_et_al_analysis/Downstream_Analysis/Chromosome_19_to_38/data/20230518_jurkat_glm_mpra_merge_temporary_38.txt"
mpra <- read.table(mpra.dir, sep="\t", header=T)
# mpra <- subset(mpra, !is.na(A.log2FC) | !is.na(B.log2FC) | !is.na(A.logPadj_BH) | !is.na(B.logPadj_BH) | !is.na(A.logPadj_BF) | !is.na(B.logPadj_BF))
mpra$A.log2FC[is.na(mpra$A.log2FC)]<- (0)
mpra$B.log2FC[is.na(mpra$B.log2FC)]<- (0)
mpra$A.logPadj_BH[is.na(mpra$A.logPadj_BH)]<- (0)
mpra$B.logPadj_BH[is.na(mpra$B.logPadj_BH)]<- (0)
mpra$A.logPadj_BF[is.na(mpra$A.logPadj_BF)]<- (0)
mpra$B.logPadj_BF[is.na(mpra$B.logPadj_BF)]<- (0)

mpra_name <- "20230518_unstimulated_jurkat_glm"
filter <- "none"

```

```{r, eval=FALSE, echo=FALSE}
# These are the different options for saving the html. They depend on the values mpra_name and filter.
if(filter == "none" & mpra_name == "20230518_unstimulated_jurkat_glm"){
  render("20231115_LOG_precision_recall_analysis.Rmd",output_file = paste0(mpra_name, "_LOG_PICS_precision_recall",".html"))
}
```

```{r}
# This html version uses the mpra data from 20230518_GM12878_emVAR_glm_mpra_merge.txt
# This html version has no filter
mpra.dir <- "/nfs/jray/screens/ALL_MPRAs/Ho_et_al_analysis/Downstream_Analysis/Chromosome_19_to_38/data/20230518_GM12878_emVAR_glm_mpra_merge_temporary_38.txt"
mpra <- read.table(mpra.dir, sep="\t", header=T)
mpra_name <- "20230518_GM12878_emVAR_glm"
filter <- "none"
```

```{r, eval=FALSE, echo=FALSE}
# These are the different options for saving the html. They depend on the values mpra_name and filter.
# This is going to be the default 
if(filter == "none" & mpra_name == "20230518_GM12878_emVAR_glm"){
  render("20231115_LOG_precision_recall_analysis.Rmd",output_file = paste0(mpra_name, "_LOG_PICS_precision_recall",".html"))
}
```

# Special Columns 

This runs a loop to find the max fold changes.
```{r max FC column}
#For each MPRA variant, find the fold change of the allele with the largest absolute fold change
# This creates a column full of NAs for the loop to fill
mpra$max_abs_fc <- NA
# This loop calculates the column of max_abs_fc. It takes the two alleles A and B and calculates which one's absolute value is larger. That larger value is the value of the new column (can be positive or negative). 
for(i in 1:nrow(mpra)){
  if(abs(mpra[i,]$A.log2FC)>abs(mpra[i,]$B.log2FC)){
    mpra[i,]$max_abs_fc<-mpra[i,]$A.log2FC
  }else{
    mpra[i,]$max_abs_fc<-mpra[i,]$B.log2FC
  }
}
```

This runs a loop to find the max fold change p values with the Benjamini Hochberg correction.
```{r max BH column}
#For each MPRA variant, find the fold change of the allele with the largest absolute fold change
# This creates a column full of NAs for the loop to fill
mpra$max.logPadj_BH<-NA
# This loop calculates the column of max_abs_fc. It takes the two alleles A and B and calculates which one's absolute value is larger. That larger value is the value of the new column (can be positive or negative). 
for(i in 1:nrow(mpra)){
  if(abs(mpra[i,]$A.logPadj_BH)>abs(mpra[i,]$B.logPadj_BH)){
    mpra[i,]$max.logPadj_BH<-mpra[i,]$A.logPadj_BH
  }else{
    mpra[i,]$max.logPadj_BH<-mpra[i,]$B.logPadj_BH
  }
}
```

This runs a loop to find the max fold change p values with the bonferroni correction.
```{r max BF column}
#For each MPRA variant, find the fold change of the allele with the largest absolute fold change
# This creates a column full of NAs for the loop to fill
mpra$max.logPadj_BF<-NA
# This loop calculates the column of max_abs_fc. It takes the two alleles A and B and calculates which one's absolute value is larger. That larger value is the value of the new column (can be positive or negative). 
for(i in 1:nrow(mpra)){
  if(abs(mpra[i,]$A.logPadj_BF)>abs(mpra[i,]$B.logPadj_BF)){
    mpra[i,]$max.logPadj_BF<-mpra[i,]$A.logPadj_BF
  }else{
    mpra[i,]$max.logPadj_BF<-mpra[i,]$B.logPadj_BF
  }
}
```


```{r max pics column}
#For each MPRA variant, find the pics for all the diseases and make max_pics the maximum of all of those values
mpra.pics <- subset(mpra, select=c(MS_pics,RA_pics,UC_pics,T1D_pics,IBD_pics,Crohns_pics,Psoriasis_pics))

# Make the NAs for pics -1
mpra.pics$MS_pics[is.na(mpra.pics$MS_pics)]<- (-1)
mpra.pics$RA_pics[is.na(mpra.pics$RA_pics)]<- (-1)
mpra.pics$UC_pics[is.na(mpra.pics$UC_pics)]<- (-1)
mpra.pics$IBD_pics[is.na(mpra.pics$IBD_pics)]<- (-1)
mpra.pics$T1D_pics[is.na(mpra.pics$T1D_pics)]<- (-1)
mpra.pics$Crohns_pics[is.na(mpra.pics$Crohns_pics)]<- (-1)
mpra.pics$Psoriasis_pics[is.na(mpra.pics$Psoriasis_pics)]<- (-1)
colnames(mpra.pics)
mpra.pics$win_pics <- colnames(mpra.pics)[apply(mpra.pics,1,which.max)] # Make a column that is called win_pics which will give me the name of the pics column with the best 

mpra.pics$max_pics <- NA # This creates a column full of NAs for the loop to fill

for(i in 1:nrow(mpra.pics)){
  if(mpra.pics[i,]$win_pics=="MS_pics"){
    mpra.pics[i,]$max_pics<-mpra.pics[i,]$MS_pics
  } # if win_pics says MS_pics then the max_pics is MS_pics
  if(mpra.pics[i,]$win_pics=="RA_pics"){
    mpra.pics[i,]$max_pics<-mpra.pics[i,]$RA_pics
  } # if win_pics says RA_pics then the max_pics is RA_pics
  if(mpra.pics[i,]$win_pics=="UC_pics"){
    mpra.pics[i,]$max_pics<-mpra.pics[i,]$UC_pics
  } # if win_pics says UC_pics then the max_pics is UC_pics
  if(mpra.pics[i,]$win_pics=="IBD_pics"){
    mpra.pics[i,]$max_pics<-mpra.pics[i,]$IBD_pics
  } # if win_pics says IBD_pics then the max_pics is IBD_pics
  if(mpra.pics[i,]$win_pics=="Crohns_pics"){
    mpra.pics[i,]$max_pics<-mpra.pics[i,]$Crohns_pics
  } # if win_pics says Crohns_pics then the max_pics is Crohns_pics
  if(mpra.pics[i,]$win_pics=="Psoriasis_pics"){
    mpra.pics[i,]$max_pics<-mpra.pics[i,]$Psoriasis_pics
  } # if win_pics says Psoriasis_pics then the max_pics is Psoriasis_pics
   if(mpra.pics[i,]$win_pics=="T1D_pics"){
    mpra.pics[i,]$max_pics<-mpra.pics[i,]$T1D_pics
  } # if win_pics says T1D_pics then the max_pics is T1D_pics
}

# Make the MPRA pics columns which are NA return to NA
mpra.pics$MS_pics[mpra.pics$MS_pics<0]<- NA
mpra.pics$RA_pics[mpra.pics$RA_pics<0]<- NA
mpra.pics$UC_pics[mpra.pics$UC_pics<0]<- NA
mpra.pics$IBD_pics[mpra.pics$IBD_pics<0]<- NA
mpra.pics$T1D_pics[mpra.pics$T1D_pics<0]<- NA
mpra.pics$Crohns_pics[mpra.pics$Crohns_pics<0]<- NA
mpra.pics$Psoriasis_pics[mpra.pics$Psoriasis_pics<0]<- NA

```

```{r, echo=FALSE}
if(filter == "none" & mpra_name == "20230518_unstimulated_jurkat_glm"){
mpra$A.log2FC[mpra$A.log2FC==0] <- NA
mpra$B.log2FC[mpra$B.log2FC==0] <- NA
mpra$A.logPadj_BH[mpra$A.logPadj_BH==0] <- NA
mpra$B.logPadj_BH[mpra$B.logPadj_BH==0]<- NA
mpra$A.logPadj_BF[mpra$A.logPadj_BF==0]<- NA
mpra$B.logPadj_BF[mpra$B.logPadj_BF==0]<- NA
}
```



```{r}
mpra$max_pics <- mpra.pics$max_pics
```

How many variants in this version of MPRA have a maximum pics value of greater than or equal to 0.5
```{r}
nrow(subset(mpra, max_pics >= 0.5))
```

# Custom axes based on the min and max values

```{r axes, eval=FALSE, echo=FALSE}
pics_true <- subset(mpra, max_pics >= 0.5)

skew_max <- round(max(pics_true$LogSkew, na.rm = TRUE),digits = 2) 
skew_min <- round(min(pics_true$LogSkew, na.rm = TRUE),digits = 2) 
skew_seq <- round(seq(0, skew_max, length = 10),digits = 2)

```




# Precision and Recall Plots

Skew heat map.
```{r correct heatmap}
skew_seq <-c(0, 0.25, 0.5, 0.75, 1, 1.25, 1.5, 1.75, 2, 3)
FDR_seq <-c(0.001, 0.01, 0.05, 0.1, 0.15, 0.2, 0.25, 0.5, 0.75, 1)
precision.skew.dat <- NULL
for( i in FDR_seq){
  for(j in skew_seq){
   mpra_subset <- subset(mpra, select=c(Skew.logFDR,LogSkew,max_abs_fc,max.logPadj_BH,max.logPadj_BF,max_pics)) # subset the mpra to only the columns we want. 
   all_true <- nrow(subset(mpra_subset, max_pics>=0.5)) #all true in that they are all likely causal according to statistical fine-mapping. This measurement is the true positives and false negatives (which are truely positive). All true goes into the denominator of precision. 
  # mpra_subset$SkewFDR<- 10^-mpra_subset$Skew.logFDR # undo the negative log transformation of the Skew log p-values and create the column SkewFDR
  # mpra_subset$Skew<- 2^mpra_subset$LogSkew # undo the log transformation of logSkew to create a column called skew
   true_positive <- nrow(subset(mpra_subset, Skew.logFDR >= i & LogSkew >= j & max_pics>=0.5)) # This variable are the true positives (TP). They are emVars which meet the specific requirements of each variable and are positive pics values (meaning they are truly significant). This is in the numerator in both precision and recall. This is variable A in the risk ratio calculation. 
   false_positive <- nrow(subset(mpra_subset, Skew.logFDR >= i & LogSkew >= j & max_pics<=0.001)) # This variable is false positives (FP). They are the non-Emvars which meet the specific requirements of the LogSkew and Skew.logFDR and have the poor pics values. This is part of the denominator in precision.This is variable B in the risk ratio calculation (except that we are using the max_pics <= 0.5)
      false_negative <- nrow(subset(mpra_subset, Skew.logFDR < i & LogSkew < j & max_pics>=0.5)) # This variable are the false negatives (FN). They are not classified as emVars which meet the specific requirements of each variable, but have positive pics values (meaning they are truly significant). This is variable C in the risk ratio calculation. 
      true_negative <- nrow(subset(mpra_subset, Skew.logFDR < i & LogSkew < j & max_pics<=0.001)) # This variable are the true negatives (TN). They are not classified as emVars which meet the specific requirements of each variable, and have poor pics values (meaning they are truly NOT significant). This is variable D in the risk ratio calculation (except that we are using the max_pics <= 0.5)
      A <- true_positive
      B <- false_positive
      C <- false_negative
      D <-  true_negative
         risk_ratio <- ((A)/(A+B)) / ((C)/((C+D))) # This is the risk ratio which is calculated as (TP / (TP + FP)) / (FN / (FN+TN)). Risk ratio is INF when False Negative is 0. Risk ratio is blank False Negative and True Negative are 0.
   precision <- true_positive/(true_positive+false_positive) # Precision is calculated by True Positives / (True Positives + False Positives). 
   recall <- true_positive/all_true # Recall is calculated as True Positives / (True Positives + False Negatives). All True is True Positives + False Negatives (which are actually true)
   precision.skew.dat <- rbind(precision.skew.dat, data.frame(i, j, precision, recall, all_true, true_positive, false_positive, false_negative, true_negative, risk_ratio, B, D)) # makes the dataframe used for the heat map
  }
}
names(precision.skew.dat) <- c("FDR", "LogSkew", "precision", "recall", "all_true", "true_positive", "false_positive", "false_negative", "true_negative", "risk_ratio", "B", "D")

ggplot(data.frame(precision.skew.dat), aes(x = as.character(FDR), y = as.character(LogSkew), fill = precision)) +
  geom_tile() + geom_text(aes(label = round(precision, digit=2))) + xlab("-log Skew FDR") + ylab("LogSkew") + 
  ggtitle("Precision heatmap for LogSkew") + scale_fill_gradient(low = "gold", high = "firebrick2") 
ggplot(data.frame(precision.skew.dat), aes(x = as.character(FDR), y = as.character(LogSkew), fill = recall)) +
  geom_tile() + geom_text(aes(label = round(recall, digit=2))) + xlab("-log Skew FDR") + ylab("LogSkew") + 
  ggtitle("Recall heatmap for LogSkew") + scale_fill_gradient(low = "gold", high = "firebrick2")
ggplot(data.frame(precision.skew.dat), aes(x = as.character(FDR), y = as.character(LogSkew), fill = risk_ratio)) +
  geom_tile() + geom_text(aes(label = round(risk_ratio, digit=2))) + xlab("-log Skew FDR") + ylab("LogSkew") + 
  ggtitle("Risk Ratio heatmap for LogSkew") + scale_fill_gradient(low = "gold", high = "firebrick2")
# Risk ratio is blank or INF when False Negative is 0
ggplot(data.frame(precision.skew.dat), aes(x = as.character(FDR), y = as.character(LogSkew), fill = true_positive)) +
  geom_tile() + geom_text(aes(label = round(true_positive, digit=2))) + xlab("-log Skew FDR") + ylab("LogSkew") + 
  ggtitle("True Positive heatmap for LogSkew") + scale_fill_gradient(low = "gold", high = "firebrick2")
ggplot(data.frame(precision.skew.dat), aes(x = as.character(FDR), y = as.character(LogSkew), fill = false_positive)) +
  geom_tile() + geom_text(aes(label = round(false_positive, digit=2))) + xlab("-log Skew FDR") + ylab("LogSkew") + 
  ggtitle("False Positive heatmap for LogSkew") + scale_fill_gradient(low = "gold", high = "firebrick2")
ggplot(data.frame(precision.skew.dat), aes(x = as.character(FDR), y = as.character(LogSkew), fill = all_true)) +
  geom_tile() + geom_text(aes(label = round(all_true, digit=2))) + xlab("-log Skew FDR") + ylab("LogSkew") + 
  ggtitle("All True heatmap for LogSkew") + scale_fill_gradient(low = "gold", high = "firebrick2")
ggplot(data.frame(precision.skew.dat), aes(x = as.character(FDR), y = as.character(LogSkew), fill = false_negative)) +
  geom_tile() + geom_text(aes(label = round(false_negative, digit=2))) + xlab("-log Skew FDR") + ylab("LogSkew") + 
  ggtitle("False Negative heatmap for LogSkew") + scale_fill_gradient(low = "gold", high = "firebrick2")
ggplot(data.frame(precision.skew.dat), aes(x = as.character(FDR), y = as.character(LogSkew), fill = true_negative)) +
  geom_tile() + geom_text(aes(label = round(true_negative, digit=2))) + xlab("-log Skew FDR") + ylab("LogSkew") + 
  ggtitle("True Negative heatmap for LogSkew") + scale_fill_gradient(low = "gold", high = "firebrick2")
```

Now attempt to create a heat map for maximum absolute fold change
```{r FC heatmap}
FC_seq <-c(0, 0.25, 0.5, 0.75, 1, 1.25, 1.5, 1.75, 2)
FDR_seq <-c(0.01, 0.25, 0.5, 0.75, 1, 1.25, 1.5, 1.75, 2, 3)
precision.FC.dat <- NULL
for( i in FDR_seq){
  for(j in FC_seq){
   mpra_subset <- subset(mpra, select=c(Skew.logFDR,LogSkew,max_abs_fc,max.logPadj_BH,max.logPadj_BF,max_pics)) # subset the mpra to only the columns we want. 
   all_true <- nrow(subset(mpra_subset, max_pics>=0.5)) #all true in that they are all likely causal according to statistical fine-mapping. This measurement is the true positives and false negatives (which are truely positive). All true goes into the denominator of precision. 
  # mpra_subset$SkewFDR<- 10^-mpra_subset$Skew.logFDR # undo the negative log transformation of the Skew log p-values and create the column SkewFDR
  # mpra_subset$FC<- 2^mpra_subset$max_abs_fc # undo the log transformation of log 2 fold change to create a column called FC
   true_positive <- nrow(subset(mpra_subset, Skew.logFDR >= i & max_abs_fc >= j & max_pics>=0.5)) # This variable are the true positives. They are emVars which meet the specific requirements of each variable and are positive pics values (meaning they are truly significant). This is in the numerator in both precision and recall. 
   false_positive <- nrow(subset(mpra_subset, Skew.logFDR >= i & max_abs_fc >= j & max_pics<=0.001)) # This variable is false positives. They are the non-Emvars which meet the specific requirements of the max_abs_fc and Skew.logFDR and have the poor pics values. This is part of the denominator in precision. 
    false_negative <- nrow(subset(mpra_subset, Skew.logFDR < i & max_abs_fc < j & max_pics>=0.5)) # This variable are the false negatives (FN). They are not classified as emVars which meet the specific requirements of each variable, but have positive pics values (meaning they are truly significant). This is variable C in the risk ratio calculation. 
    true_negative <- nrow(subset(mpra_subset, Skew.logFDR < i & max_abs_fc < j & max_pics<=0.001)) # This variable are the true negatives (TN). They are not classified as emVars which meet the specific requirements of each variable, and have poor pics values (meaning they are truly NOT significant). This is variable D in the risk ratio calculation (except that we are using the max_pics <= 0.5)
      A <- true_positive
      B <- false_positive
      C <- false_negative
      D <- true_negative
         risk_ratio <- ((A)/(A+B)) / ((C)/((C+D))) # This is the risk ratio which is calculated as (TP / (TP + FP)) / (FN / (FN+TN)). Risk ratio is INF when False Negative is 0. Risk ratio is blank False Negative and True Negative are 0.
   precision <- true_positive/(true_positive+false_positive) # Precision is calculated by True Positives / (True Positives + False Positives). 
   recall <- true_positive/all_true # Recall is calculated as True Positives / (True Positives + False Negatives). All True is True Positives + False Negatives (which are actually true)
   precision.FC.dat <- rbind(precision.FC.dat, data.frame(i, j, precision, recall, all_true, true_positive, false_positive, risk_ratio)) # makes the dataframe used for the heat map
  }
}
names(precision.FC.dat) <- c("FDR", "max_abs_fc", "precision", "recall", "all_true", "true_positive", "false_positive", "risk_ratio")

ggplot(data.frame(precision.FC.dat), aes(x = as.character(FDR), y = as.character(max_abs_fc), fill = precision)) +
  geom_tile() + geom_text(aes(label = round(precision, digit=2))) + xlab("-log Skew FDR") + ylab("max_abs_fc") + 
  ggtitle("Precision heatmap for max_abs_fc") + scale_fill_gradient(low = "gold", high = "firebrick2") 
ggplot(data.frame(precision.FC.dat), aes(x = as.character(FDR), y = as.character(max_abs_fc), fill = recall)) +
  geom_tile() + geom_text(aes(label = round(recall, digit=2))) + xlab("-log Skew FDR") + ylab("max_abs_fc") + 
  ggtitle("Recall heatmap for max_abs_fc") + scale_fill_gradient(low = "gold", high = "firebrick2")
ggplot(data.frame(precision.FC.dat), aes(x = as.character(FDR), y = as.character(max_abs_fc), fill = risk_ratio)) +
  geom_tile() + geom_text(aes(label = round(risk_ratio, digit=2))) + xlab("-log Skew FDR") + ylab("max_abs_fc") + 
  ggtitle("Risk Ratio heatmap for max_abs_fc") + scale_fill_gradient(low = "gold", high = "firebrick2")
ggplot(data.frame(precision.FC.dat), aes(x = as.character(FDR), y = as.character(max_abs_fc), fill = true_positive)) +
  geom_tile() + geom_text(aes(label = round(true_positive, digit=2))) + xlab("-log Skew FDR") + ylab("max_abs_fc") + 
  ggtitle("True Positive heatmap for max_abs_fc") + scale_fill_gradient(low = "gold", high = "firebrick2")
```

Heat map for max.logPadj_BH 

```{r BH heatmap}
BH_seq <- c(0.25, 0.5, 0.75, 1, 1.25, 1.5, 1.75, 2, 2.25, 2.5)
FDR_seq <-c(0.01, 0.25, 0.5, 0.75, 1, 1.25, 1.5, 1.75, 2, 3)
precision.BH.dat <- NULL
for( i in FDR_seq){
  for(j in BH_seq){
   mpra_subset <- subset(mpra, select=c(Skew.logFDR,LogSkew,max_abs_fc,max.logPadj_BH,max.logPadj_BF,max_pics)) # subset the mpra to only the columns we want. 
   all_true <- nrow(subset(mpra_subset, max_pics>=0.5)) #all true in that they are all likely causal according to statistical fine-mapping. This measurement is the true positives and false negatives (which are truely positive). All true goes into the denominator of precision. 
  # mpra_subset$SkewFDR<- 10^-mpra_subset$Skew.logFDR # undo the negative log transformation of the Skew log p-values and create the column SkewFDR
  # mpra_subset$BH<- 10^-mpra_subset$max.logPadj_BH # undo the log transformation of log 2 fold change to create a column called BH
   true_positive <- nrow(subset(mpra_subset, Skew.logFDR >= i & max.logPadj_BH >= j & max_pics>=0.5)) # This variable are the true positives. They are emVars which meet the specific requirements of each variable and are positive pics values (meaning they are truly significant). This is in the numerator in both precision and recall. 
   false_positive <- nrow(subset(mpra_subset, Skew.logFDR >= i & max.logPadj_BH >= j & max_pics<=0.001)) # This variable is false positives. They are the non-Emvars which meet the specific requirements of the max.logPadj_BH and Skew.logFDR and have the poor pics values. This is part of the denominator in precision.  
    false_negative <- nrow(subset(mpra_subset, Skew.logFDR < i & max.logPadj_BH < j & max_pics>=0.5)) # This variable are the false negatives (FN). They are not classified as emVars which meet the specific requirements of each variable, but have positive pics values (meaning they are truly significant). This is variable C in the risk ratio calculation. 
    true_negative <- nrow(subset(mpra_subset, Skew.logFDR < i & max.logPadj_BH < j & max_pics<=0.001)) # This variable are the true negatives (TN). They are not classified as emVars which meet the specific requirements of each variable, and have poor pics values (meaning they are truly NOT significant). This is variable D in the risk ratio calculation (except that we are using the max_pics <= 0.5 for risk ratio)
      A <- true_positive
      B <- false_positive
      C <- false_negative
      D <- true_negative 
         risk_ratio <- ((A)/(A+B)) / ((C)/((C+D))) # This is the risk ratio which is calculated as (TP / (TP + FP)) / (FN / (FN+TN)). Risk ratio is INF when False Negative is 0. Risk ratio is blank False Negative and True Negative are 0. 
   precision <- true_positive/(true_positive+false_positive) # Precision is calculated by True Positives / (True Positives + False Positives). 
   recall <- true_positive/all_true # Recall is calculated as True Positives / (True Positives + False Negatives). All True is True Positives + False Negatives (which are actually true)
   precision.BH.dat <- rbind(precision.BH.dat, data.frame(i, j, precision, recall, all_true, true_positive, false_positive, risk_ratio)) # makes the dataframe used for the heat map
  }
}
names(precision.BH.dat) <- c("FDR", "max.logPadj_BH", "precision", "recall", "all_true", "true_positive", "false_positive", "risk_ratio")

ggplot(data.frame(precision.BH.dat), aes(x = as.character(FDR), y = as.character(max.logPadj_BH), fill = precision)) +
  geom_tile() + geom_text(aes(label = round(precision, digit=2))) + xlab("-log Skew FDR") + ylab("max.logPadj_BH") + 
  ggtitle("Precision heatmap for max.logPadj_BH") + scale_fill_gradient(low = "gold", high = "firebrick2") 
ggplot(data.frame(precision.BH.dat), aes(x = as.character(FDR), y = as.character(max.logPadj_BH), fill = recall)) +
  geom_tile() + geom_text(aes(label = round(recall, digit=2))) + xlab("-log Skew FDR") + ylab("max.logPadj_BH") + 
  ggtitle("Recall heatmap for max.logPadj_BH") + scale_fill_gradient(low = "gold", high = "firebrick2")
ggplot(data.frame(precision.BH.dat), aes(x = as.character(FDR), y = as.character(max.logPadj_BH), fill = risk_ratio)) +
  geom_tile() + geom_text(aes(label = round(risk_ratio, digit=2))) + xlab("-log Skew FDR") + ylab("max.logPadj_BH") + 
  ggtitle("Risk Ratio heatmap for max.logPadj_BH") + scale_fill_gradient(low = "gold", high = "firebrick2")
ggplot(data.frame(precision.BH.dat), aes(x = as.character(FDR), y = as.character(max.logPadj_BH), fill = true_positive)) +
  geom_tile() + geom_text(aes(label = round(true_positive, digit=2))) + xlab("-log Skew FDR") + ylab("max.logPadj_BH") + 
  ggtitle("True Positive heatmap for max.logPadj_BH") + scale_fill_gradient(low = "gold", high = "firebrick2")
```

Heat map for max.logPadj_BF
```{r BF Heatmap}
BF_seq <- c(0.25, 0.5, 0.75, 1, 1.25, 1.5, 1.75, 2, 2.25, 2.5)
FDR_seq <-c(0.01, 0.25, 0.5, 0.75, 1, 1.25, 1.5, 1.75, 2, 3)
precision.BF.dat <- NULL
for( i in FDR_seq){
  for(j in BF_seq){
   mpra_subset <- subset(mpra, select=c(Skew.logFDR,LogSkew,max_abs_fc,max.logPadj_BH,max.logPadj_BF,max_pics)) # subset the mpra to only the columns we want. 
   all_true <- nrow(subset(mpra_subset, max_pics>=0.5)) #all true in that they are all likely causal according to statistical fine-mapping. This measurement is the true positives and false negatives (which are truely positive). All true goes into the denominator of precision. 
  # mpra_subset$SkewFDR<- 10^-mpra_subset$Skew.logFDR # undo the negative log transformation of the Skew log p-values and create the column SkewFDR
  # mpra_subset$BF<- 10^-mpra_subset$max.logPadj_BF # undo the log transformation of log 2 fold change to create a column called BF
   true_positive <- nrow(subset(mpra_subset, Skew.logFDR >= i & max.logPadj_BF >= j & max_pics>=0.5)) # This variable are the true positives. They are emVars which meet the specific requirements of each variable and are positive pics values (meaning they are truly significant). This is in the numerator in both precision and recall. 
   false_positive <- nrow(subset(mpra_subset, Skew.logFDR >= i & max.logPadj_BF >= j & max_pics<=0.001)) # This variable is false positives. They are the non-Emvars which meet the specific requirements of the max.logPadj_BF and Skew.logFDR and have the poor pics values. This is part of the denominator in precision (except that we are using the max_pics <= 0.5 for risk ratio)
    false_negative <- nrow(subset(mpra_subset, Skew.logFDR < i & max.logPadj_BF < j & max_pics>=0.5)) # This variable are the false negatives (FN). They are not classified as emVars which meet the specific requirements of each variable, but have positive pics values (meaning they are truly significant). This is variable C in the risk ratio calculation. 
    true_negative <- nrow(subset(mpra_subset, Skew.logFDR < i & max.logPadj_BF < j & max_pics<=0.001)) # This variable are the true negatives (TN). They are not classified as emVars which meet the specific requirements of each variable, and have poor pics values (meaning they are truly NOT significant). This is variable D in the risk ratio calculation (except that we are using the max_pics <= 0.5 for risk ratio)
      A <- true_positive
      B <- false_positive
      C <- false_negative
      D <- true_negative
         risk_ratio <- ((A)/(A+B)) / ((C)/((C+D))) # This is the risk ratio which is calculated as (TP / (TP + FP)) / (FN / (FN+TN)). Risk ratio is INF when False Negative is 0. Risk ratio is blank False Negative and True Negative are 0. 
   precision <- true_positive/(true_positive+false_positive) # Precision is calculated by True Positives / (True Positives + False Positives). 
   recall <- true_positive/all_true # Recall is calculated as True Positives / (True Positives + False Negatives). All True is True Positives + False Negatives (which are actually true)
   precision.BF.dat <- rbind(precision.BF.dat, data.frame(i, j, precision, recall, all_true, true_positive, false_positive, risk_ratio)) # makes the dataframe used for the heat map
  }
}
names(precision.BF.dat) <- c("FDR", "max.logPadj_BF", "precision", "recall", "all_true", "true_positive", "false_positive", "risk_ratio")

ggplot(data.frame(precision.BF.dat), aes(x = as.character(FDR), y = as.character(max.logPadj_BF), fill = precision)) +
  geom_tile() + geom_text(aes(label = round(precision, digit=2))) + xlab("-log Skew FDR") + ylab("max.logPadj_BF") + 
  ggtitle("Precision heatmap for max.logPadj_BF") + scale_fill_gradient(low = "gold", high = "firebrick2") 
ggplot(data.frame(precision.BF.dat), aes(x = as.character(FDR), y = as.character(max.logPadj_BF), fill = recall)) +
  geom_tile() + geom_text(aes(label = round(recall, digit=2))) + xlab("-log Skew FDR") + ylab("max.logPadj_BF") + 
  ggtitle("Recall heatmap for max.logPadj_BF") + scale_fill_gradient(low = "gold", high = "firebrick2")
ggplot(data.frame(precision.BF.dat), aes(x = as.character(FDR), y = as.character(max.logPadj_BF), fill = risk_ratio)) +
  geom_tile() + geom_text(aes(label = round(risk_ratio, digit=2))) + xlab("-log Skew FDR") + ylab("max.logPadj_BF") + 
  ggtitle("Risk Ratio heatmap for max.logPadj_BF") + scale_fill_gradient(low = "gold", high = "firebrick2")
ggplot(data.frame(precision.BF.dat), aes(x = as.character(FDR), y = as.character(max.logPadj_BF), fill = true_positive)) +
  geom_tile() + geom_text(aes(label = round(true_positive, digit=2))) + xlab("-log Skew FDR") + ylab("max.logPadj_BF") + 
  ggtitle("True Positive heatmap for max.logPadj_BF") + scale_fill_gradient(low = "gold", high = "firebrick2")
```







